### DISCUSSION TITLE
Get vaultwarden to "proxy" or pull icos from a pre-cache? ie. not accessing internet

### DISCUSSION THREAD

**Discussion by hvisage** [dani-garcia/vaultwarden#2108] [Ideas]
Good day,

Yes I'm pedantic lately.

I don't want my Vaultwarden to connect out to the BBI, and thus blocks all outgoing connections except those that is absolutely needed for operations of server. 

Then enters the favicon.ico lookups for the Vault, which makes things spin-and-spin-and-spin appears to be "broken".

Has anybody thought of implementing a cache or a method to pre-fill the icons? - ideally it'll be a place where icons are pre-tested for anything bad like too big etc.

**Comment by hvisage** [comment]
Okay, I think I found the "solution" (for me atleast ;) )

setup `/etc/rsyncd.conf`
```
[ic]
 uid = bitwardenrs
 use chroot = true
 path = /opt/bitwarden_rs/data/icon_cache
 hosts allow = 10.100.255.10
 read only = no
```

from the cache fetcher:
```
cd icon_cache
rsync -av 10.22.228.2::ic/ --include=\*miss --exclude=\* . --remove-source-files
rsync -v 10.22.228.2::ic/ --include=\*png --exclude=\* -a .
```
I want to remove the .miss files but not the rest
```
ls |sed -e 's/\(.*\).png\(.miss\)*/\1/'|sort -u| awk '{print "rm "$1"*miss ; wget https://www.google.com/s2/favicons\?domain\=https://"$1" -O "$1".png "}'|bash
rsync -va . 10.22.228.2::ic/
```

Thank you Google for the Favicon "service" ;)

**Reply by BlackDex** [reply]
Yea that would be the proper solution if you do not have internet access available for the Vaultwarden instance it self.
I would also recommend to set the `ICON_CACHE_TTL` to `0` so that they are never being refreshed by Vaultwarden.

**Reply by BlackDex** [reply]
Also, you can use other favicon sources btw if someone doesn't like google to be used for this.

- The official Bitwarden services: https://icons.bitwarden.net/${DOMAIN}/icon.png
- DuckDuckGO: https://icons.duckduckgo.com/ip3/${DOMAIN}.ico
